### PR TITLE
chore: add yaml schema

### PR DESCRIPTION
**Pull request by 1yefuwang1**
Add yaml schema for better yaml file editing experience

### CODE REVIEW

File: packages/fx-core/resource/yaml-schema.json [hunk 1]
```diff
@@ -0,0 +1,1495 @@
...
+      "type": "string",
+      "description": "path to the environment folder"
+    },
+    "version": {
+      "type": "string",
+      "description": "The version of the yaml file schema"
+    },
+    "provision": {
+      "$ref": "#/definitions/lifeCycleArray",
+      "description": "Triggered by `teamsfx provision`"
```
Comment (timngmsft) [inline packages/fx-core/resource/yaml-schema.json:20]: I think the word "Called" is better then "Triggered"?

File: packages/fx-core/resource/yaml-schema.json [hunk 2]
```diff
@@ -0,0 +1,1495 @@
...
+  "type": "object",
+  "additionalProperties": false,
+  "properties": {
+    "projectId": {
+      "type": "string",
+      "description": "projectId used for telemetry"
+    },
+    "environmentFolderPath": {
+      "type": "string",
+      "description": "path to the environment folder"
```
Comment (timngmsft) [inline packages/fx-core/resource/yaml-schema.json:12]: Some descriptions (like on line 16) start with capitalized words, and others do not. Which is it?

File: packages/fx-core/resource/yaml-schema.json [hunk 3]
```diff
@@ -0,0 +1,1495 @@
...
+      "description": "Create Azure Active Directory application and client secret (optional). Please refer to https://aka.ms/teamsfx-actions/aadapp-create for more details.",
+      "required": ["uses", "writeToEnvironmentFile"],
+      "properties": {
+        "name": {
+          "type": "string",
+          "description": "An optional name of this action"
+        },
+        "env": {
+          "type": "object",
+          "description": "Define environment variables",
```
Comment (timngmsft) [inline packages/fx-core/resource/yaml-schema.json:78]: What is this? Is this inputs or outputs?

File: packages/fx-core/resource/yaml-schema.json [hunk 4]
```diff
@@ -0,0 +1,1495 @@
...
+          { "$ref": "#/definitions/script" }
+        ]
+      }
+    },
+    "aadAppCreateBase": {
+      "type": "object",
+      "description": "Create Azure Active Directory application and client secret (optional). Please refer to https://aka.ms/teamsfx-actions/aadapp-create for more details.",
+      "required": ["uses", "writeToEnvironmentFile"],
+      "properties": {
+        "name": {
```
Comment (timngmsft) [inline packages/fx-core/resource/yaml-schema.json:72]: How do we use this? I see its optional and never generated in our templates ... should we just remove this?

File: packages/fx-core/resource/yaml-schema.json [hunk 5]
```diff
@@ -0,0 +1,1495 @@
...
+        "env": {
+          "type": "object",
+          "description": "Define environment variables",
+          "additionalProperties": {
+            "type": "string"
+          }
+        },
+        "uses": {
+          "type": "string",
+          "description": "Create Azure Active Directory application when the environment variable that stores clientId is empty. Also create client secret for the application when generateClientSecret parameter is true and the environment variable that stores clientSecret is empty. When creating new Azure Active Directory application, this action generates clientId, objectId, tenantId, authority and authorityHost. When creating new client secret, this action generates clientSecret. Please refer to https://aka.ms/teamsfx-actions/aadapp-create for more details.",
```
Comment (timngmsft) [inline packages/fx-core/resource/yaml-schema.json:85]: Where is this environment variable for `clientId`? Is it TTK environment or shell environment? What if its not empty?

**Review by timngmsft — CHANGES_REQUESTED** [review]

File: packages/fx-core/resource/yaml-schema.json [hunk 6]
```diff
@@ -0,0 +1,1495 @@
...
+                "additionalProperties": false,
+                "required": ["deploymentName", "path"],
+                "properties": {
+                  "deploymentName": {
+                    "type": "string",
+                    "description": "The name of ARM deployment"
+                  },
+                  "path": {
+                    "type": "string",
+                    "description": "Relative path to ARM template. Both bicep and json format are supported."
```
Comment (timngmsft) [inline packages/fx-core/resource/yaml-schema.json:307]: Can you do this clean up in the entire file - but Bicep should be capitalized, and JSON should be all caps.

File: packages/fx-core/resource/yaml-schema.json [hunk 7]
```diff
@@ -0,0 +1,1495 @@
...
+              "description": "Generate the final Azure Active Directory application manifest used to update Azure Active Directory application to this path."
+            }
+          }
+        }
+      }
+    },
+    "armDeploy": {
+      "type": "object",
+      "additionalProperties": false,
+      "description": "Execute ARM deployment. Refer to https://aka.ms/teamsfx-actions/arm-deploy for more details on the naming convertion rule.",
```
Comment (timngmsft) [inline packages/fx-core/resource/yaml-schema.json:256]: How about "Create Azure resources using the referenced Bicep / JSON files."

File: packages/fx-core/resource/yaml-schema.json [hunk 8]
```diff
@@ -0,0 +1,1495 @@
...
+        "env": {
+          "type": "object",
+          "description": "Define environment variables",
+          "additionalProperties": {
+            "type": "string"
+          }
+        },
+        "uses": {
+          "type": "string",
+          "description": "Execute ARM deployment. Every bicep output will be persisted in current environment's .env file with certain naming convertion. Refer to https://aka.ms/teamsfx-actions/arm-deploy for more details on the naming convertion rule.",
```
Comment (timngmsft) [inline packages/fx-core/resource/yaml-schema.json:272]: "Create Azure resources using the referenced Bicep / JSON files. Outputs from Bicep / /JSON will be persisted in the current Teams Toolkit environment. Refer to ..."

File: packages/fx-core/resource/yaml-schema.json [hunk 9]
```diff
@@ -0,0 +1,1495 @@
...
+              }
+            }
+          }
+        }
+      }
+    },
+    "azureStorageEnableStaticWebsite": {
+      "type": "object",
+      "additionalProperties": false,
+      "description": "Enable static website config for Azure Storage. Please refer to https://aka.ms/teamsfx-actions/azure-storage-enable-static-website for more details.",
```
Comment (timngmsft) [inline packages/fx-core/resource/yaml-schema.json:323]: I noticed some places you have "Refer to ..." and other places you have "Please refer to ...". We should use "Refer to ..." to be consistent.

**Review by timngmsft — CHANGES_REQUESTED** [review]

File: packages/fx-core/resource/yaml-schema.json (hunk 4 above)
Comment (1yefuwang1) [reply packages/fx-core/resource/yaml-schema.json:72]: Some templates/samples uses "name" field to give each `cli/npmRunCommand` a human readable name. It's usesful when there're more than one `cli/npmRunCommand` actions defined for difference purposes. For example, one is for building tab another is for building bot. Without a human readable name, a user cannot easily tell them apart in both yaml file and summary info.

Comment (1yefuwang1) [reply packages/fx-core/resource/yaml-schema.json:72]: An example is here https://github.com/OfficeDev/TeamsFx-Samples/blob/dev/hello-world-teams-tab-and-outlook-add-in/teamsapp.yml#L52

File: packages/fx-core/resource/yaml-schema.json (hunk 3 above)
Comment (1yefuwang1) [reply packages/fx-core/resource/yaml-schema.json:78]: It's inputs. It's for defining environemnt variables in the context of this action, which is especially useful for `cli/npmRunScript`, `script` and actions for deployment.

File: packages/fx-core/resource/yaml-schema.json (hunk 5 above)
Comment (1yefuwang1) [reply packages/fx-core/resource/yaml-schema.json:85]: This is the `aadAppCreateBase` which is referenced by `aadAppCreateWithSecret` and `aadAppCreateWithoutSecret` as a base "class" to share common fields(see "allOf" in L93).
Schema of clientId is defined in L122 and L182.
If it's not empty, a new AAD app will be created and the new AAD appid will be stored in the envrionment variable specified by clientID. @blackchoey

Comment (1yefuwang1) [reply packages/fx-core/resource/yaml-schema.json:85]: aadAppCreate is of type either `aadAppCreateWithSecret` or `aadAppCreateWithoutSecret`. We implement it this way to provider better intellisense.

File: packages/fx-core/resource/yaml-schema.json (hunk 9 above)
Comment (1yefuwang1) [reply packages/fx-core/resource/yaml-schema.json:323]: All "Please refer to" has been replaced with "Refer to".

File: packages/fx-core/resource/yaml-schema.json (hunk 2 above)
Comment (1yefuwang1) [reply packages/fx-core/resource/yaml-schema.json:12]: All descriptions have been changed to start with capitalized words.

File: packages/fx-core/resource/yaml-schema.json (hunk 1 above)
Comment (1yefuwang1) [reply packages/fx-core/resource/yaml-schema.json]: Done

File: packages/fx-core/resource/yaml-schema.json (hunk 4 above)
Comment (1yefuwang1) [reply packages/fx-core/resource/yaml-schema.json:72]: @timngmsft I'm resolving this conversation. Please let me know if there's further questions.

**Review by timngmsft — APPROVED** [review]

**Review by jayzhang — DISMISSED** [review]

File: packages/fx-core/resource/package.nls.json [hunk 10]
```diff
@@ -1021,6 +1021,7 @@
   "driver.botFramework.progressBar.createOrUpdateBot": "Creating or updating bot registration.",
   "driver.generateAppsettings.description": "Add or update environment variables to appsettings file.",
   "error.yaml.InvalidYamlSchemaError": "Unable to parse yaml file: %s",
+  "error.yaml.InvalidYamlSchemaErrorWithReason": "Unable to parse yaml file: %s. Reason: ",
```
Comment (sffamily) [inline packages/fx-core/resource/package.nls.json:1024]: after "Reason" there is no any string for it, is it supposed to be like this?

Comment (1yefuwang1) [reply packages/fx-core/resource/package.nls.json]: Nice catch. if there is no %s, the extra parameter string will be appended to the end.
image

That's why I missed it.

**Review by jayzhang — DISMISSED** [review]

**Review by sffamily — APPROVED** [review]

**Review by hund030 — APPROVED** [review]

**Review by hund030 — APPROVED** [review]

**Review by jayzhang — APPROVED** [review]

**Review by sffamily — APPROVED** [review]

**Pull request merged**